File: rings.cs
using System;
using System.Collections.Generic;

class Program
{
    // Метод для ввода строки с кольцами
    static string Input()
    {
        Console.Write("Введите строку: ");
        return Console.ReadLine();
    }

    // Метод для вывода результата
    static void Output(int result)
    {
        Console.WriteLine($"Количество стержней с 3 цветами: {result}");
    }

    static void Main()
    {
        string rings = Input();
        
        // Словарь, где ключ - стержень, а значение - множество для хранения уникальных цветов
        Dictionary<int, HashSet<char>> rodColors = new Dictionary<int, HashSet<char>>();
        
        // Инициализация словаря для 10 стержней (0-9)
        for (int i = 0; i < 10; i++)
        {
            rodColors[i] = new HashSet<char>();
        }

        // Обработка строки колец (цвет и номер стержня)
        for (int i = 0; i < rings.Length; i += 2)
        {
            char color = rings[i];         // Цвет кольца
            int rod = rings[i + 1] - '0'; // Номер стержня (преобразование char в int)
            rodColors[rod].Add(color);    // Добавляем цвет к стержню
        }

        int result = 0; // Счётчик стержней с 3 цветами

        // Подсчёт стержней с тремя разными цветами
        foreach (var rod in rodColors)
        {
            if (rod.Value.Count == 3)
            {
                result++;
            }
        }

        Output(result);
    }
}

File: rings.py
def input_rings():
    rings = input("Введите строку: ")
    return rings

def output_result(result):
    print(f"Количество стержней с 3 цветами: {result}")

rings = input_rings()

# Словарь для хранения уникальных цветов на каждом стержне
rod_colors = {i: set() for i in range(10)}

# Проходим по строке с кольцами и цветами
for i in range(0, len(rings), 2):
    color = rings[i] # Цвет кольца
    rod = int(rings[i + 1]) # Стержень кольца
    rod_colors[rod].add(color) # Добавляем цвет в соответствующий стержень

result = 0 # Количество стержней с 3 цветами

# Считаем количество стержней с 3 цветами
for rod in rod_colors:
    if len(rod_colors[rod]) == 3:
        result += 1

output_result(result)